[C#] add dictionaries exercise (#1497)
exercism/v3: languages/csharp/exercises/concept/dictionaries/.docs/introduction.md

> A dictionary is a collection of elements where each element comprises a key and value such that if a key is passed to a method of the ditionary its associated value is returned. It has the same role as maps or associative arrays do in other languages.

Reviewer: I really like this introduction! Concise and explained very well.

A dictionary can be created as follows:

```csharp

> new Dictionary<int, string>{{1, "One"}, {2, "Two"}};

Reviewer: I was wondering if maybe having the element on a separate line would be helpful here? It also seems to be the format commonly used when initializing a dictionary (each element on a separate line).

```suggestion
new Dictionary<int, string>
{
    {1, "One"},
    {2, "Two"}
};
```

// 1 => "One", 2 => "Two"
```

Note that the key and value types are part of the definition of the dictionary.

> Entries can be added to the dictionary as follows:
> 
> ```csharp
> var numbers = new Dictionary<int, string>();
> numbers.Add(3, "Three");
> numbers.Add(4, "Four");
> // 3 => "Three", 4 => "Four"
> ```
> 
> To remove a key-value pair simply pass the key to the `Remove` method.

Reviewer: This example does not introduce any new syntax, right? As such, could you remove the example? The text could become something like:

```suggestion
Once constructed, entries can be added or removed from a dictionary using its built-in methods.
```

I do wonder if maybe it is useful to add the `new Dictionary<int, string>()` way of creating a dictionary to the above example that uses the object initializer syntax.

> A dictionary lookup takes the following form:
> 
> ```csharp
> var numbers = new Dictionary<int, string>{{1, "One"}, {2, "Two"}};
> numbers[2];
> // "Two"
> ```
> 
> This is known as an [indexer][indexer-properties].
> ]
> 
> A value in the dictionary can be replaced as follows:
> 
> ```csharp
> var numbers = new Dictionary<int, string>{{1, "One"}, {2, "Two"}};
> numbers[2] = "Deux";
> // {{1, "One"}, {2, "Deux"}}
> ```

Reviewer: Maybe these examples can be combined?

```suggestion
Retrieving or updating values in a dictionary is done by indexing into the dictionary using a key:

```csharp
var numbers = new Dictionary<int, string>
{
   {1, "One"},
   {2, "Two"}
};

// Set the value of the element with key 2 to "Deux"
numbers[2] = "Deux";

// Get the value of the element with key 2
numbers[2];
// "Deux"
```

This could perhaps even be combined with the first example where a dictionary is declared. It might be interesting to look at the [Arrays exercise's introduction.md document](https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/arrays/.docs/introduction.md).

You can test if a value exists in the dictionary with:

```csharp
var dict = new Dictionary<string, string>{/*...*/};
dict.ContainsKey("some key that exists");
// => true

> ```

Reviewer: This does not introduce any new syntax, right? As such, could you remove this from the introduction?

> You can enumerate the following from the dictionary, keys, values, key-value pairs:
> 
> ```csharp
> var dict = new Dictionary<string, int>{/*...*/};
> 
> Dictionary<string, int>.KeyCollection keys = dict.Keys;
> 
> Dictionary<string, int>.ValueCollection values = dict.Values;
> 
> IEnumerable<KeyValuePair<string,int>> enumerable = dict;
> ```

Reviewer: This does not introduce any new syntax, right? As such, could you remove the example from the introduction? Maybe we can rephrase to something like:

```suggestion
Enumerating over a dictionary will enumerate over its key/value pairs. Dictionaries also have properties that allow enumerating over its keys or values.
```

Or something like that?
